Implement hardware connectivity confirmation screens

darkwing opened the pull request:
Designs found here:  https://www.figma.com/file/6Px3A6vVFGy1hYbEUXi343/Hardware-Wallets?node-id=2%3A2020

## ToDo:

- [x] Implement form fields in popover
- [x] Implement popover padding and other styling
- [x] Remove hardcoded debug values
- [x] Implement polling of connectivity status
- [x] Remove LedgerInstructionsField
- [x] Update Ledger keyring library

## Manual Testing
1.  Pull down this PR
2. In `node_modules/@metamask/eth-ledger-bridge-keyring/index.js`, change the `BRIDGE_URL` to `'https://darkwing.github.io/eth-ledger-bridge-keyring'`; that address includes the `gh-pages` portion of this project (https://github.com/MetaMask/eth-ledger-bridge-keyring/pull/147).  Also, comment out the following check for origin:

```
/*
      if (origin !== this._getOrigin()) {
        return false
      }
*/
```

3. Connect a Ledger address, then try to do a basic send on testnet -- you will see the connectivity messaging and functionality!

## Considerations
1.  Dropdown mode vs. full screen mode
2. The different Ledger connectivity options

darkwing committed:
Implement hardware connectivity confirmation screens

darkwing committed:
Dropdown progress

darkwing committed:
Progress

darkwing committed:
Remove path config from popover, lint

darkwing committed:
Add hardware connectivity functionality

darkwing committed:
Remove unused component

darkwing committed:
Wire up actual calls to the controller to poll for ledger connectivity

darkwing committed:
Prevent navigation event during connection

jpuri commented:
Hey @darkwing : seems there is no activity on the PR from sometime.

darkwing commented:
@jpuri This PR is basically done; the majority of the work needs to be done in the `eth-ledger-bridge-keyring` repo.

darkwing committed:
Cleanup console logs

darkwing commented on ui/pages/confirm-transaction-base/hardware-connectivity/hardware-connectivity-content.js:
```diff
@@ -0,0 +1,62 @@
+            <ol>
+              <li>Plug in your Ledger</li>
+              <li>Enter your passcode</li>
+              <li>Open the Ethereum app</li>
```
ToDo:  Use localization for those steps.

KOOLMOLL replied:
ToDo: Use localization for those steps.

darkwing commented on ui/pages/confirm-transaction-base/hardware-connectivity/hardware-connectivity-content.js:
```diff
@@ -0,0 +1,62 @@
+          >
+            {t('hardwareWalletConnectivityHelperHeading', [deviceName])}
+          </Typography>
+          <img src="./images/ledger.svg" />
```
TODO:  Add `alt=""` here.

darkwing commented on ui/pages/confirm-transaction-base/confirm-transaction-base.component.js:
```diff
@@ -153,8 +157,46 @@ export default class ConfirmTransactionBase extends Component {
+    this.pollLedgerReady();
+    const intervalId = setInterval(() => {
+      this.pollLedgerReady();
+    }, 2000);
```
TODO: Change `2000` to a constant.

darkwing committed:
Hide connection popover as soon as connection happens

darkwing committed:
Localize Ledger steps

darkwing committed:
Remove logs, update state to not show advanced upon connection

darkwing committed:
Use constants

darkwing committed:
Fix incorrect keyring localization

darkwing committed:
Assume hardware ready upon connection

darkwing committed:
Fix colors per Thomas' request

darkwing committed:
Add icons

tmashuang committed:
Fixes ci, lint, and unit test issues for hardware connectivity. (#13748)

* Remove unusual locales
* Lint ignore line for todo function
* Reduce length check for settings content row, but add LedgerTransportDropdown component check

darkwing committed:
Remove console.logs

danjm commented on ui/components/app/confirm-page-container/confirm-page-container-content/confirm-page-container-content.component.js:
```diff
@@ -128,6 +129,7 @@ export default class ConfirmPageContainerContent extends Component {
       nativeCurrency,
       networkName,
       showBuyModal,
+      showingHardwareConnectionContents,
```
Might be good to generalize this variable name. Within this component, this variable just determines whether we should show `ConfirmPageContainerSummary` or not. So we could change this to `hideConfirmPageContainerSummary`

darkwing replied:
I prefer not to name things with a negative word; also, `showingHardwareConnectionContents` is akin to `showBuyModal`, representing whether or not we are showing the popover.  Maybe you'd prefer `showingHardwareConnectionPopover`?

danjm replied:
I was mostly thinking to remove the reference to the hardware wallet. In the context of this file, the hardware wallet is irrelevant, the boolean property just indicates whether or not to show `ConfirmPageContainerSummary`... the parent of this component can decide to hide or show that for whatever reason it chooses (not just because of the hardware wallet being connected)

danjm commented on ui/pages/confirm-transaction-base/confirm-transaction-base.component.js:
```diff
@@ -979,6 +1074,11 @@ export default class ConfirmTransactionBase extends Component {
     this._removeBeforeUnload();
+
+    const { pollingIntervalId } = this.state;
+    if (pollingIntervalId) {
```
Should we also clear the interval on a window `beforeunload` event?

danjm reviewed (changes requested):
This looks pretty good! Left a couple of small comments. Should we also get a design review from @holantonela ?

darkwing committed:
Clear interval upon page unload

danjm reviewed (changes requested):
This looks good to me. We just need to get the ledger keyring library updated

segun commented:
On this branch, I see ledger is not connected even though my device is connected

darkwing committed:
Add close button to connect modal

darkwing committed:
Remove unused LedgerInstructionField component

darkwing committed:
Broaden language

darkwing committed:
Rename prop to be more descriptive

darkwing committed:
Turn connection link to button

darkwing committed:
Change name of checkLedgerReady to more generic name

darkwing committed:
Don't show connect button when connected

darkwing committed:
Allow connecting in the popup to work

AlexJupiter commented:
@holantonela @darkwing I've created a metrics ticket for this experience here: https://github.com/MetaMask/metamask-extension/issues/14014

darkwing committed:
Remove duplicate classname declaration

darkwing committed:
Fix success text color

darkwing committed:
Fix lint

darkwing committed:
Fix merge issues

darkwing commented on ui/components/ui/actionable-message/index.scss:
```diff
@@ -114,6 +114,15 @@
+    background-color: #eaf8ed;
+
+    .actionable-message__message {
+      color: var(--Black-100);
```
TODO: Update these colors

darkwing commented on ui/pages/create-account/connect-hardware/connect-hardware-path.js:
```diff
@@ -0,0 +1,27 @@
+
+ConnectHardwarePathSelector.propTypes = {
+  className: PropTypes.string,
+  onChange: PropTypes.func,
```
Add `isRequired` here.

darkwing committed:
Fix colors and prop

digiwand commented on ui/pages/settings/advanced-tab/ledger-transport-dropdown.js:
```diff
@@ -0,0 +1,120 @@
+      <div className="settings-page__content-item">
+        <div className="settings-page__content-item-col">
+          <Dropdown
+            id="select-ledger-transport-type"
```
I see that this id name was copied over verbatim, but might I suggest we change this to match the file name? This is to promote more robust code standards and avoid naming collisions.

darkwing replied:
Cool with me!

digiwand reviewed (commented):
Took a wing at reviewing this code though it's my first time looking at the hardware logic. Left some suggestions and saw no blockers! 🎉 

It's also my first time running the development code with the ledger connectivity. I checked out this branch and successfully sent a transaction with this code. I'm going to step through some of the code in action before submitting a final approval

digiwand commented on ui/pages/confirm-transaction-base/hardware-connectivity/hardware-connectivity-content.js:
```diff
@@ -0,0 +1,77 @@
+            <i
+              onClick={onClose}
+              onKeyUp={(event) => {
+                if (event.key === 'Enter') {
```
Q: Will more tests and storybook logic be added to support the new changes?

darkwing commented on ui/pages/confirm-transaction-base/confirm-transaction-base.component.js:
```diff
@@ -157,8 +162,68 @@ export default class ConfirmTransactionBase extends Component {
+
+    // Don't set off multiple calls to checkDeviceReady
+    if (pollingIntervalId !== null) {
+      return undefined;
```
At one point we required a return value or linting would fail, but maybe that's now gone.

darkwing commented on ui/pages/confirm-transaction-base/confirm-transaction-base.component.js:
```diff
@@ -157,8 +162,68 @@ export default class ConfirmTransactionBase extends Component {
+      return undefined;
+    }
+
+    let hardwareIsReady = true;
```
Yes -- if something is somehow faulty with our code, we don't want to prevent users from being able to do whatever they'd like.

darkwing commented on ui/pages/confirm-transaction-base/hardware-connectivity/hardware-connectivity-message.js:
```diff
@@ -0,0 +1,50 @@
+
+export default function HardwareConnectivityMessage({
+  connected = false,
+  onClick = undefined,
```
Good shout!

darkwing commented on ui/pages/confirm-transaction-base/confirm-transaction-base.component.js:
```diff
@@ -577,6 +644,33 @@ export default class ConfirmTransactionBase extends Component {
+      return (
+        <div className="confirm-page-container-content__details">
+          <HardwareConnectivityContent
+            deviceName="Ledger"
```
Good shout!

darkwing commented on ui/pages/confirm-transaction-base/hardware-connectivity/hardware-connectivity-content.js:
```diff
@@ -0,0 +1,77 @@
+  const t = useI18nContext();
+
+  switch (deviceName) {
+    case 'Ledger':
```
Yes!!!!

darkwing committed:
Address feedback

darkwing committed:
Fix lint

digiwand reviewed (commented):
a couple of more comments coming in after stepping through the code and some more inspection

darkwing commented on ui/pages/create-account/connect-hardware/connect-hardware-path.js:
```diff
@@ -0,0 +1,27 @@
+import React from 'react';
```
Oh, we changed direction after this was created, so not used anymore.  Really good find!

darkwing committed:
Remove unused file

darkwing commented on ui/pages/settings/advanced-tab/ledger-transport-dropdown.js:
```diff
@@ -0,0 +1,120 @@
+import React, { useState } from 'react';
```
Done!  Great catch!

darkwing committed:
Move transport dropdown to component library

darkwing committed:
Disable ledger connectivity UI for Firefox

darkwing committed:
Add storybook for LedgerTransportDropdown

brad-decker reviewed (dismissed):
Tested with my ledger over the weekend using the library update and this is working very well.

darkwing committed:
Fix outdated link

darkwing committed:
Rename variable

darkwing committed:
EXPERIMENTAL FOR LINUX: Launch into full screen

danjm commented on ui/store/actions.js:
```diff
@@ -406,6 +407,8 @@ export function connectHardware(deviceName, page, hdPath, t) {
+    const isPopup = getEnvironmentType() === ENVIRONMENT_TYPE_POPUP;
+
     dispatch(
       showLoadingIndication(`Looking for your ${capitalize(deviceName)}...`),
```
`Looking for your ${capitalize(deviceName)}...` needs to be localized

darkwing replied:
This isn't new -- it's an existing issue.  We don't have `this.context.t` available in actions.  I believe there's an issue for this open already.

danjm reviewed (dismissed):
Only one comment from me. This looks good! I tested the solution for when webhid is disconnected, and it works.

darkwing committed:
Let the SEND screen direct when to stop and start keyring connection polling

digiwand committed:
HardwareConnectivity: support dark mode ledger svg
this is awkward... the ledger.svg file disappeared from my IDE.
might need to deleted in another commit

digiwand commented on ui/pages/confirm-transaction-base/confirm-transaction-base.component.js:
```diff
@@ -981,6 +1085,7 @@ export default class ConfirmTransactionBase extends Component {
   componentWillUnmount() {
     this._beforeUnloadForGasPolling();
     this._removeBeforeUnload();
+    this._clearPollingInterval();
```
@darkwing bump - just want to double-check this won't cause any memory leaks. If we're unsure, we can add the remove listener just in case.

```suggestion
    window.removeEventListener('beforeunload', () => this._clearPollingInterval);
    this._clearPollingInterval();
```

I see `this._removeBeforeUnload();` is being called above, but it's removing the event with a different callback
```
  _removeBeforeUnload = () => {
    window.removeEventListener('beforeunload', this._beforeUnloadForGasPolling);
  };
```

digiwand reviewed (commented):
LGTM and appears to be working well on Chrome on my macOS! I left some optional suggestions. 

The one thing holding me back from approving is this event listener that I believe we want to remove on unmount: 
https://github.com/MetaMask/metamask-extension/pull/12725#discussion_r870849004

darkwing committed:
Update ui/pages/confirm-transaction-base/hardware-connectivity/index.scss

Co-authored-by: Ariella Vu <20778143+digiwand@users.noreply.github.com>

darkwing committed:
Address feedback

darkwing commented:
Addressed feedback!

digiwand reviewed (dismissed):
LGTM! 🙌🏼

danjm commented:
@darkwing This needs to be rebased onto develop to get all tests passing.

digiwand reviewed (approved):
🙌

brad-decker commented:
@darkwing @danjm @hilvmason @kevinghim this seems like important work that we failed to get landed and is now seriously out of date. @darkwing i'm sorry that happened to you. Should we put this PR into sprint queue to prioritize review / rebase time for it?

darkwing commented:
@brad-decker I don't think so.  The priority is MV3, and this PR could conflict with the work the team is doing on hardware wallet stuff for MV3.  Also, this PR and its dependency heavily rely on `setInterval` and `setTimeout`, which will no longer work with MV3, so it's best that the MV3 work gets done first.

brad-decker commented:
@darkwing sounds good, but lets plan to revisit this conversation and have the review / rebase time dedicated as part of a future sprint after MV3. In the meantime, i'll convert to draft.

holantonela commented:
We now have dedicated extension development capacity at the Key Management team, so perhaps @darkwing and @HowardBraham can work together on this after MV3 priors get updated. @AlexJupiter will coordinate it.
